inductiva/localization/localization.py

```diff
@@ -0,0 +1,143 @@
+"""Localization module that provides support for multiple languages."""
+from typing import Optional
+from threading import RLock
+import pathlib
+import logging
+import locale
+import json
+
+
+_INDUCTIVA_DEFAULT_LANG = "en"
+
+
+_logger = logging.getLogger(__name__)
+
+
+class Translator:
+    """Translation provider with string rendering support.
+    
+    Attributes:
+        TRANSLATIONS_DIR (pathlib.Path): Path to the directory where the
+            translation files are stored.
+    """
+    TRANSLATIONS_DIR = pathname = pathlib.Path(__file__).parent / "translations"
+
+    class _Template(str):
+        """Custom string class to allow for the use of the % operator"""
+
+        def __mod__(self, other):
+            """
+            Allow for the use of the % operator to format the string
+            with the str.format method.
+            """
+            if isinstance(other, dict):
+                return self.format(**other)
+            elif isinstance(other, tuple):
+                return self.format(*other)
+            return self.format(other)
+        
+        def __str__(self) -> str:
+            # hack to allow logging to work with this class.
+            # logging uses msg = str(self.msg) in getMessage() before applying
+            # the % operator. Here we return the object itself to avoid route
+            # the string formatting to the __mod__ method
+            return self
```

Comment:
I am not sure that this is necessary. It is my understanding that this is just being used to format string. But we can use simpler python commands. The following works:

```python
>>> with open("../pt.json", "r") as f: translation = json.load(f)
... 
>>> translation
{'upgrade-available': ["A versão {0} da biblioteca 'inductiva' está disponível (a versão atualmente instalada é {1}).", 'Considere atualizar para a versão mais recente usando:', '  pip install --upgrade inductiva'], 'failed-update-check': ["Não foi possível verificar se há atualizações para a biblioteca 'inductiva'.", 'O seguinte erro ocorreu:\n  {0}']}
>>> translation['upgrade-available'][0].format("2", "3")
"A versão 2 da biblioteca 'inductiva' está disponível (a versão atualmente instalada é 3)."
```